Add `litgpt evaluate` command

rasbt opened the pull request:
This adds the `litgpt evaluate` command along with a wrapper function that allows someone to evaluate models without needing to convert or copy anything manually. I.e., it's just 

```bash
litgpt finetune lora \
  --checkpoint_dir checkpoints/microsoft/phi-2 \
  --out_dir lora_model
```

```bash
litgpt evaluate \
  --checkpoint_dir lora_model/final \
  --out_dir evaluate_model/ \
```

Fixes #1168

rasbt commented:
One question is how to add the tests @carmocca because I assume it's very expensive. One could probably test

it on the smallest phi-2 model and on a small number of examples e.g.

```
litgpt evaluate \
  --checkpoint_dir checkpoints/EleutherAI/pythia-14m/ \
  --out_dir out/ \
  --repo_id EleutherAI/pythia-14m \
  --limit 10 \
  --tasks "hellaswag"
```

Still, I don't know if the CI will be able to handle it. I am also not sure if there's a good way to mock this.

carmocca commented:
@rasbt I think you should be able to update the tests in https://github.com/Lightning-AI/litgpt/blob/main/tests/test_lm_eval_harness.py#L30. In fact we might want to remove the existing eval implementation

rasbt commented:
Thanks, I hope the test I added suffices

awaelchli commented on litgpt/scripts/evaluate.py:
Wondering what's the reason for this being in scripts? Is it because we are essentially wrapping another library and therefore it is not part of the core files? cc @carmocca

carmocca replied:
I think it's fine to put this under `eval/`. I would also completely replace the current `eval/` at the top level. As long as the users interact with this through CLI, they won't care how eval is actually happening under the hood

awaelchli commented on litgpt/scripts/evaluate.py:
```diff
@@ -0,0 +1,116 @@
+        limit: Limit on number of examples per task.
+        seed: Random seed.
+        save_filepath: The file where the results will be saved. 
+          Saves to `out_dir`/results.json by default.
```
```suggestion
            Saves to `out_dir/results.json` by default.
```

awaelchli commented on litgpt/scripts/evaluate.py:
```diff
@@ -0,0 +1,116 @@
+
+    from lm_eval import evaluator
+
+    if checkpoint_dir is None:
+        raise ValueError("Provide a checkpoint_dir argument.")
+    if out_dir is None:
+        raise ValueError("Provide a checkpoint_dir argument.")
+    if repo_id is None:
+        raise ValueError("Provide a repo_id argument.")   
```
If these are required, why not mark them as required?
Also could we derive the `repo_id` from the config file so we don't have to pass it in?

rasbt replied:
That's a great idea!

awaelchli commented on litgpt/scripts/evaluate.py:
```diff
@@ -0,0 +1,116 @@
+
+    checkpoint_dir, out_dir = Path(checkpoint_dir), Path(out_dir)
+
+    if save_filepath is None:
+        save_filepath = "results.json"
+        save_filepath = out_dir / Path(save_filepath)
+    else:
+        save_filepath = Path(save_filepath)
```
```suggestion
    save_filepath = out_dir / Path("results.json") if save_filepath is None else Path(save_filepath)
```

awaelchli commented on litgpt/scripts/evaluate.py:
```diff
@@ -0,0 +1,116 @@
+    checkpoint_dir, out_dir = Path(checkpoint_dir), Path(out_dir)
+
+    if save_filepath is None:
+        save_filepath = "results.json"
+        save_filepath = out_dir / Path(save_filepath)
```
```suggestion
        save_filepath = out_dir / Path("results.json")
```

awaelchli commented on litgpt/scripts/evaluate.py:
```diff
@@ -0,0 +1,116 @@
+            print(make_table(results, "groups"))
+
+    json_result = json.dumps(
+            results, indent=2, ensure_ascii=False
```
```suggestion
        results, indent=2, ensure_ascii=False
```

rasbt commented:
Thanks for all the feedback. These were all great suggestions!

The latest updates include

1. Make `checkpoint_dir` a required argument
2. Set default `output_dir` as `checkpoint_dir/evaluate`
3. Moved `evalute.py` out of `litgpt/scripts`
4. Automatically infer the `repo_id` from the config file 

Question 1:  Should we make `output_dir` the same as `checkpoint_dir` by default (`evaluate` will save the converted model + results.json)?

Question 2: I moved it to `litgpt/eval/evaluate.py`. Since it's a single module, we could also just make it `litgpt/evaluate.py` and remove the `eval subfolder. Wdyt?

carmocca commented on tests/test_evaluate.py:
```diff
@@ -0,0 +1,62 @@
+    match="Loading a dataset cached in a LocalFileSystem is not supported",
+)
+def test_run_eval(tmp_path, float_like):
+    repo_id = "EleutherAI/pythia-14m"
+    download_from_hub(repo_id=repo_id, checkpoint_dir=tmp_path)
+
+    checkpoint_path = Path(tmp_path) / Path(repo_id)
+
+    convert_lit_checkpoint(checkpoint_dir=checkpoint_path, output_dir=checkpoint_path)
+    safe_safetensors(out_dir=checkpoint_path, repo_id=repo_id)
+
+    eval_tasks = "coqa,hellaswag"
+    results = evaluator.simple_evaluate(
+        model="hf",
+        model_args=f"pretrained={checkpoint_path}",
+        tasks=eval_tasks.split(","),
+        limit=2,
+        device="cpu"
+    )
+
+    save_path = checkpoint_path/"results.json"
+    prepare_results(results, save_path, print_results=False)
```
Can you call the evaluate script from the test instead of repeating its internal steps?

rasbt replied:
Sounds good. Let's revisit later after addressing the other issue regarding saving the converted model

carmocca commented on tests/test_lm_eval_harness.py:
```diff
@@ -1,95 +0,0 @@
-    assert (output_folder / "results.json").read_text() == '{"foo": "test"}'
-
-
-def test_cli():
```
Can you add back this test for the new CLI command? Also parameterize it as done for the same test in lora

rasbt replied:
fixed

carmocca commented on litgpt/eval/evaluate.py:
```diff
@@ -0,0 +1,114 @@
+    model = AutoModel.from_pretrained(
+        repo_id, state_dict=state_dict
+    )
+    model.save_pretrained(out_dir)
```
Do we really need to save it here? Does the lm eval harness not support taking a HF model instance instead?

rasbt replied:
You mean a model instance generated via `[convert_lit_checkpoint.py](https://github.com/Lightning-AI/litgpt/blob/main/litgpt/scripts/convert_lit_checkpoint.py)`? This usually never worked with the harness. I think @awaelchli also tried that before without success.

carmocca replied:
I mean using the `model` instance you have in L18 which according to their docs can be passed in https://github.com/EleutherAI/lm-evaluation-harness/blob/main/lm_eval/evaluator.py#L61-L62

The advantage being that you skip the step of saving another copy of the weights

rasbt replied:
I see. I tried that (see current PR) but having issues with this. Now resulting in a 

```
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 235/235 [00:00<00:00, 816.08it/s]
2024-03-27:17:55:45,864 INFO     [task.py:395] Building contexts for truthfulqa_mc2 on rank 0...
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 817/817 [00:00<00:00, 981.97it/s]
2024-03-27:17:55:46,750 INFO     [task.py:395] Building contexts for gsm8k on rank 0...
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [00:05<00:00, 229.79it/s]
2024-03-27:17:55:52,516 INFO     [task.py:395] Building contexts for hellaswag on rank 0...
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10042/10042 [00:03<00:00, 3273.56it/s]
2024-03-27:17:55:56,577 INFO     [evaluator.py:362] Running loglikelihood requests
Running loglikelihood requests:   0%|                                                                                                                                                  | 0/109439 [00:00<?, ?it/s]Traceback (most recent call last):
  File "/home/zeus/miniconda3/envs/cloudspace/bin/litgpt", line 8, in <module>
    sys.exit(main())
  File "/teamspace/studios/this_studio/litgpt/litgpt/__main__.py", line 131, in main
    fn(**kwargs)
  File "/teamspace/studios/this_studio/litgpt/litgpt/eval/evaluate.py", line 100, in convert_and_evaluate
    results = evaluator.simple_evaluate(
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/lm_eval/utils.py", line 288, in _wrapper
    return fn(*args, **kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/lm_eval/evaluator.py", line 234, in simple_evaluate
    results = evaluate(
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/lm_eval/utils.py", line 288, in _wrapper
    return fn(*args, **kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/lm_eval/evaluator.py", line 373, in evaluate
    resps = getattr(lm, reqtype)(cloned_reqs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/lm_eval/api/model.py", line 329, in loglikelihood
    return self._loglikelihood_tokens(new_reqs, disable_tqdm=disable_tqdm)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/lm_eval/models/huggingface.py", line 1033, in _loglikelihood_tokens
    self._model_call(batched_inps, **call_kwargs), dim=-1
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/lm_eval/models/huggingface.py", line 748, in _model_call
    return self.model(inps).logits
AttributeError: 'BaseModelOutputWithPast' object has no attribute 'logits'
```

Not sure how to fix that, I find it hard to figure out what I am missing. Unless it's something obvious you maybe know how to fix, I suggest reverting back to the state before and save the converted model even though it's wasteful. We can then open an issue and reinvestigate that later and see if we find a way to evaluate for the HF object directly. This would not change the API or anything, so I would say let's leave that as an optional thing?

awaelchli commented on litgpt/eval/evaluate.py:
```diff
@@ -0,0 +1,114 @@
+
+    copy_config_files(source_dir=checkpoint_dir, out_dir=out_dir)
+
+    if not skip_conversion:
+        convert_lit_checkpoint(checkpoint_dir=checkpoint_dir, output_dir=out_dir)
+        safe_safetensors(out_dir, repo_id)
```
I noticed that when running multiple times, the `convert_lit_checkpoint` would always run even if the model is already converted to safetensors, but the safe_safetensors gets skipped. I think we probably want this to be consistent, so either we skip if the output dir already contains the converted model, or we always overwrite it.

rasbt replied:
Good point. I think this should be addressed now.

awaelchli commented on litgpt/eval/evaluate.py:
```diff
@@ -0,0 +1,114 @@
+        torch_random_seed=seed,
+    )
+
+    print("results", results)
```
Note that this prints a huge wall of text to the terminal (all responses to all tasks).

rasbt replied:
Ah yes! Actually I noticed that we don't need to print the results dictionary explicitly. The `evaluator.simple_evaluate` already prints the summary table internally. I modified this to only have it print that one, and the detailed output JSON is saved as the `results.json`.

awaelchli commented on tutorials/evaluation.md:
Let's also include the `--batch_size` argument in the example commands below? IMO this should almost always be set to a high number when evaluating on many tasks.

rasbt commented on tutorials/evaluation.md:
```diff
@@ -9,59 +9,71 @@ You can evaluate LitGPT using [EleutherAI's lm-eval](https://github.com/Eleuther
-### Evaluating LoRA-finetuned LLMs
+> [!TIP]
+> By default, `ligpt evaluate` will evaluate a model on all Open LM Leaderboard tasks, which corresponds
+to the setting `--tasks "hellaswag,gsm8k,truthfulqa_mc2,mmlu,winogrande,arc_challenge"`. 
```
Looks like this was a bad splitting via the Note box on GitHub. Annoying, but "hellaswag,gsm8k,truthfulqa_mc2,mmlu,winogrande,arc_challenge" needs to be a on a separate line to avoid this.

rasbt commented on tutorials/evaluation.md:
````diff
@@ -9,59 +9,71 @@ You can evaluate LitGPT using [EleutherAI's lm-eval](https://github.com/Eleuther
 ```
 
-A list of supported tasks can be found [here](https://github.com/EleutherAI/lm-evaluation-harness/blob/master/docs/task_table.md).
+Please note that the `litgpt eval` command run an internal model conversion. 
````
fixed

rasbt commented on tutorials/evaluation.md:
````diff
@@ -9,59 +9,71 @@ You can evaluate LitGPT using [EleutherAI's lm-eval](https://github.com/Eleuther
-  It is available as `truthfulqa_mc`.
+&nbsp;
+
+```
````
fixed

rasbt commented on tests/test_evaluate.py:
```diff
@@ -0,0 +1,62 @@
+)
+def test_run_eval(tmp_path, float_like):
+    repo_id = "EleutherAI/pythia-14m"
+    download_from_hub(repo_id=repo_id, checkpoint_dir=tmp_path)
```
Same as above. Let's revisit later after addressing the other issue regarding saving the converted model

rasbt commented:
I reverted it back to the previous version which worked. Sure, it might save the possibly redundant .safetensors intermediate file, but I think we need to get this PR merged as soon as possible so users can evaluate models. We can see if we can avoid this temporary file in the future as an efficiency optimization. It would not affect the user API so there shouldn't be a problem deferring this to the future.

rasbt commented:
Phew, finally got the test to work. It should be ready to review (and hopefully merge). If you have a moment, I'd appreciate your feedback @awaelchli & @carmocca

awaelchli commented on tutorials/images/0_to_litgpt/pretrain.webp:
isn't this file referenced in the 0 to litgpt doc?

rasbt replied:
Thanks for mentioning this. I must have incorrectly copied the file. I updated the image here to include the evaluate API command: 

<img width="286" alt="Screenshot 2024-04-02 at 4 21 20 PM" src="https://github.com/Lightning-AI/litgpt/assets/5618407/0b37900c-85f2-4896-95d6-ecc6e716ca27">

<br>
<br>

<img width="410" alt="Screenshot 2024-04-02 at 4 21 23 PM" src="https://github.com/Lightning-AI/litgpt/assets/5618407/3a69a651-c3fd-4095-9e07-8ac9fb00517d">
a9-b264e4b615c9">

carmocca commented on tutorials/evaluation.md:
```diff
@@ -9,59 +9,95 @@ You can evaluate LitGPT using [EleutherAI's lm-eval](https://github.com/Eleuther
-## FAQs
+> [!TIP]
+> By default, `ligpt evaluate` will evaluate a model on 3 tasks
+> to the setting `--tasks 
+> "hellaswag,truthfulqa_mc2,mmlu"`. 
```
newlines are not rendered in TIPs

```suggestion
> to the setting `--tasks "hellaswag,truthfulqa_mc2,mmlu"`. 
```

carmocca commented on tutorials/evaluation.md:
```diff
@@ -9,59 +9,95 @@ You can evaluate LitGPT using [EleutherAI's lm-eval](https://github.com/Eleuther
-The above command can be used to evaluate models that are saved via a single checkpoint file. This includes downloaded checkpoints and base models finetuned via the full and adapter finetuning scripts.
+Please note that the `litgpt evaluate` command run an internal model conversion. 
+This is only necessary the first time you want to evaluate a model. To skip the conversion, 
+when you want to evaluate a model a second time, you can pass the `--skip_conversion true` argument:
```
This says skip_conversion but the code implements force_conversion

rasbt replied:
Ah yes, this was changed based on Adrian's suggestion. We know don't run the conversion redundantly anymore. Users don't really need to know about `--force_conversion` in most cases, so I removed that from the docs.

carmocca replied:
Should we remove the argument from the script too then?

rasbt replied:
I find it useful to have as an option. E.g. if your model changed and you want to evaluate the updated model you can force the conversion. The alternative is to manually delete the files.

carmocca reviewed (commented):
I'm trying this out locally but It fails for me on:

```python
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/carmocca/git/lit-stablelm/litgpt/__main__.py", line 135, in <module>
    main()
  File "/home/carmocca/git/lit-stablelm/litgpt/__main__.py", line 131, in main
    fn(**kwargs)
  File "/home/carmocca/git/lit-stablelm/litgpt/eval/evaluate.py", line 99, in convert_and_evaluate
    results = evaluator.simple_evaluate(
  File "/home/carmocca/git/nightly-venv/lib/python3.10/site-packages/lm_eval/utils.py", line 288, in _wrapper
    return fn(*args, **kwargs)
  File "/home/carmocca/git/nightly-venv/lib/python3.10/site-packages/lm_eval/evaluator.py", line 192, in simple_evaluate
    task_dict = get_task_dict(tasks, task_manager)
  File "/home/carmocca/git/nightly-venv/lib/python3.10/site-packages/lm_eval/tasks/__init__.py", line 420, in get_task_dict
    task_name_from_string_dict = task_manager.load_task_or_group(
  File "/home/carmocca/git/nightly-venv/lib/python3.10/site-packages/lm_eval/tasks/__init__.py", line 270, in load_task_or_group
    collections.ChainMap(*map(self._load_individual_task_or_group, task_list))
  File "/home/carmocca/git/nightly-venv/lib/python3.10/site-packages/lm_eval/tasks/__init__.py", line 161, in _load_individual_task_or_group
    return load_task(task_config, task=name_or_config, group=parent_name)
  File "/home/carmocca/git/nightly-venv/lib/python3.10/site-packages/lm_eval/tasks/__init__.py", line 150, in load_task
    task_object = ConfigurableTask(config=config)
  File "/home/carmocca/git/nightly-venv/lib/python3.10/site-packages/lm_eval/api/task.py", line 782, in __init__
    self.download(self.config.dataset_kwargs)
  File "/home/carmocca/git/nightly-venv/lib/python3.10/site-packages/lm_eval/api/task.py", line 871, in download
    self.dataset = datasets.load_dataset(
  File "/home/carmocca/git/nightly-venv/lib/python3.10/site-packages/datasets/load.py", line 2556, in load_dataset
    builder_instance = load_dataset_builder(
  File "/home/carmocca/git/nightly-venv/lib/python3.10/site-packages/datasets/load.py", line 2265, in load_dataset_builder
    builder_instance: DatasetBuilder = builder_cls(
  File "/home/carmocca/git/nightly-venv/lib/python3.10/site-packages/datasets/builder.py", line 371, in __init__
    self.config, self.config_id = self._create_builder_config(
  File "/home/carmocca/git/nightly-venv/lib/python3.10/site-packages/datasets/builder.py", line 620, in _create_builder_config
    builder_config._resolve_data_files(
  File "/home/carmocca/git/nightly-venv/lib/python3.10/site-packages/datasets/builder.py", line 211, in _resolve_data_files
    self.data_files = self.data_files.resolve(base_path, download_config)
  File "/home/carmocca/git/nightly-venv/lib/python3.10/site-packages/datasets/data_files.py", line 799, in resolve
    out[key] = data_files_patterns_list.resolve(base_path, download_config)
  File "/home/carmocca/git/nightly-venv/lib/python3.10/site-packages/datasets/data_files.py", line 752, in resolve
    resolve_pattern(
  File "/home/carmocca/git/nightly-venv/lib/python3.10/site-packages/datasets/data_files.py", line 393, in resolve_pattern
    raise FileNotFoundError(error_msg)
FileNotFoundError: Unable to find 'hf://datasets/math_qa@fafb9f7ee5b9ec4da9499f9c4177a4c91389f2d6/default/train/0000.parquet' with any supported extension ['.csv', '.tsv', '.json', '.jsonl', '.parquet', '.geoparquet', '.gpq', '.arrow', '.txt', '.tar', '.blp', '.bmp', '.dib', '.bufr', '.cur', '.pcx', '.dcx', '.dds', '.ps', '.eps', '.fit', '.fits', '.fli', '.flc', '.ftc', '.ftu', '.gbr', '.gif', '.grib', '.h5', '.hdf', '.png', '.apng', '.jp2', '.j2k', '.jpc', '.jpf', '.jpx', '.j2c', '.icns', '.ico', '.im', '.iim', '.tif', '.tiff', '.jfif', '.jpe', '.jpg', '.jpeg', '.mpg', '.mpeg', '.msp', '.pcd', '.pxr', '.pbm', '.pgm', '.ppm', '.pnm', '.psd', '.bw', '.rgb', '.rgba', '.sgi', '.ras', '.tga', '.icb', '.vda', '.vst', '.webp', '.wmf', '.emf', '.xbm', '.xpm', '.BLP', '.BMP', '.DIB', '.BUFR', '.CUR', '.PCX', '.DCX', '.DDS', '.PS', '.EPS', '.FIT', '.FITS', '.FLI', '.FLC', '.FTC', '.FTU', '.GBR', '.GIF', '.GRIB', '.H5', '.HDF', '.PNG', '.APNG', '.JP2', '.J2K', '.JPC', '.JPF', '.JPX', '.J2C', '.ICNS', '.ICO', '.IM', '.IIM', '.TIF', '.TIFF', '.JFIF', '.JPE', '.JPG', '.JPEG', '.MPG', '.MPEG', '.MSP', '.PCD', '.PXR', '.PBM', '.PGM', '.PPM', '.PNM', '.PSD', '.BW', '.RGB', '.RGBA', '.SGI', '.RAS', '.TGA', '.ICB', '.VDA', '.VST', '.WEBP', '.WMF', '.EMF', '.XBM', '.XPM', '.aiff', '.au', '.avr', '.caf', '.flac', '.htk', '.svx', '.mat4', '.mat5', '.mpc2k', '.ogg', '.paf', '.pvf', '.raw', '.rf64', '.sd2', '.sds', '.ircam', '.voc', '.w64', '.wav', '.nist', '.wavex', '.wve', '.xi', '.mp3', '.opus', '.AIFF', '.AU', '.AVR', '.CAF', '.FLAC', '.HTK', '.SVX', '.MAT4', '.MAT5', '.MPC2K', '.OGG', '.PAF', '.PVF', '.RAW', '.RF64', '.SD2', '.SDS', '.IRCAM', '.VOC', '.W64', '.WAV', '.NIST', '.WAVEX', '.WVE', '.XI', '.MP3', '.OPUS', '.zip']
```

I have the same version as in CI. Are you able to run the test?

rasbt commented on tests/test_evaluate.py:
```diff
@@ -62,8 +53,13 @@ def test_evaluate_script(tmp_path, monkeypatch):
-def test_cli(fake_checkpoint_dir):
-    cli_path = Path(__file__).parent.parent / "litgpt" / "eval" / "evaluate.py"
-    output = subprocess.check_output([sys.executable, cli_path, "-h"])
+@pytest.mark.parametrize("mode", ["file", "entrypoint"])
```
Thanks a lot, I was struggling with this part.

rasbt commented:
> I'm trying this out locally but It fails for me on:

Hm, weird, it works fine for me locally. Let me try in a fresh environment, maybe it's an `lm_eval` version issue

<img width="792" alt="Screenshot 2024-04-03 at 10 26 58 AM" src="https://github.com/Lightning-AI/litgpt/assets/5618407/f0efbba9-7513-4403-aadf-dc0f23536cdb">

rasbt commented:
I totally messed up the rebasing, sry

carmocca commented:
Do you have a backup of the state? It seems like you overwrote most of the previous changes. If not, I should have one. I can force push it again

I would suggest not using rebase inside PRs since it's easier to merge and the commits will be squashed anyways when the PR lands

rasbt commented:
If you have a backup, that would be safer and nice. I don't know what happened, have done this successfully several times before, but I must have totally messed something up

carmocca commented:
Done. You'll need to redo any of the last commits you did before your force merge. I think this was only a "cleanup evaluation docs" removing one section

rasbt commented:
> Done. You'll need to redo any of the last commits you did before your force merge. I think this was only a "cleanup evaluation docs" removing one section

Thanks a lot!! And done!!

Actually, I rewrote that part regading `--force conversion` in the docs, because if users don't know about it, there can be a common gotcha where people update a checkpoint but find that it has the same evaluation performance as before due to the litgpt temp files.

carmocca commented on litgpt/eval/evaluate.py:
```diff
@@ -0,0 +1,115 @@
+from litgpt.utils import CLI, copy_config_files
+
+
+def safe_safetensors(out_dir, repo_id):
```
```suggestion
def save_safetensors(out_dir, repo_id):
```

carmocca commented on litgpt/eval/evaluate.py:
```diff
@@ -0,0 +1,115 @@
+
+    safetensors_path = out_dir / "model.safetensors"
+    if not safetensors_path.exists() or force_conversion:
+        safe_safetensors(out_dir, repo_id)
```
```suggestion
        save_safetensors(out_dir, repo_id)
```
